### ISSUE TITLE
Timeouts getting multiple Let's Encrypt certificates using TransIP DNS provider

### ISSUE THREAD

**Issue by ronaldevers** [containous/traefik#4346]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I'm running a bunch of private and one or two publicly available services using `docker-compose`. Traefik is one of the containers and I'm trying to get it to request Let's Encrypt certificates for each container. I'm using labels in the compose file to direct traefik, basically just an `enable=true` with a host rule on each container. I'm using `onHostRule=true`, the DNS-01 challenge and TransIP as a provider.

### What did you expect to see?

A certificate is requested successfully for each docker container based on its host rule (docker compose labels: `traefik.frontend.rule=Host:thing.ch10.nl`).

### What did you see instead?

Only one of the certificate requests succeeds, the rest times out. Traefik logs that it is waiting for DNS propagation for each container/domain but when I look in the TransIP DNS control panel, I can only see an acme challenge TXT record for 1 domain. So the timeouts are not because of problems with DNS propagation but because the records are not correctly set in the DNS zone.

I think the problem is that the TransIP DNS provider in Lego only has a `SetDNSEntries` method that saves an entire DNS zone. When Traefik tries to request 5 different certificates, it ends up fetching the current zone details and `SetDNSEntries` in quick succession. I think there is a race condition between getting the zone info and the `SetDNSEntries` which updates an entire zone where the calls to `SetDNSEntries` end up overwriting TXT records of concurrently running challenges.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.6
Codename:     maroilles
Go version:   go1.11.3
Built:        2018-12-14_06:43:37AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
logLevel = "INFO"
defaultEntryPoints = ["http", "https"]
[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
    entryPoint = "https"
  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]
[acme]
email = "ronald+traefik@ch10.nl"
entryPoint = "https"
storage = "acme.json"
acmeLogging = true
onHostRule = true
caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  [acme.dnsChallenge]
  resolvers = ["1.1.1.1:53"]
  provider = "transip"
[api]
[ping]
[docker]
domain = "t.ch10.nl"
exposedByDefault = false
```

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)

This log output shows that multiple challenges are tried but only the one (docs.*) succeeds.

```
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Starting provider *acme.Provider {\"Email\":\"ronald+traefik@ch10.nl\",\"ACMELogging\":true,\"CAServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"Storage\":\"acme.json\",\"EntryPoint\":\"https\",\"KeyType\":\"\",\"OnHostRule\":true,\"OnDemand\":false,\"DNSChallenge\":{\"Provider\":\"transip\",\"DelayBeforeCheck\":0,\"Resolvers\":[\"1.1.1.1:53\"],\"DisablePropagationCheck\":false},\"HTTPChallenge\":null,\"TLSChallenge\":null,\"Domains\":null,\"Store\":{}}"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Testing certificate renew..."
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :80"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :443"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :8080"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :80"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :443"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="Server configuration reloaded on :8080"
traefik      | time="2018-12-26T14:38:11Z" level=info msg="The key type is empty. Use default key type 4096."
traefik      | time="2018-12-26T14:38:14Z" level=info msg=Register...
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] acme: Registering account for ronald+traefik@ch10.nl"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="Validating FQDN authority with DNS using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] acme: Obtaining bundled SAN certificate"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] acme: Obtaining bundled SAN certificate"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] acme: Obtaining bundled SAN certificate"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] [ip.t.ch10.nl] acme: Obtaining bundled SAN certificate"
traefik      | time="2018-12-26T14:38:14Z" level=info msg="legolog: [INFO] [portainer.t.ch10.nl] acme: Obtaining bundled SAN certificate"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/tKBsFpS5LZS_LlrFXkZZmmXhKBqwV_z6PoE5cGbx74A"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] acme: Preparing to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/Tby4Ndt_OtdnlcUQ_2hJhPpr50l7wTBlH_UJLyG_ufo"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] acme: Could not find solver for: http-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] acme: Preparing to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [portainer.t.ch10.nl] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/j4ylozCFVovI12Emt6FkeNfpSSg6-sohhhnfPYJzZc4"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [portainer.t.ch10.nl] acme: Preparing to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/2SxxYm4PlNrcRPmUShWoj1gKDqHct6kcXFzYxKq2NSA"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] acme: Could not find solver for: http-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] acme: Preparing to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [ip.t.ch10.nl] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/ZEnAYzoshNp1RyMlFATAXEB7EiFiuc8aOpoDUXGyrGQ"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [ip.t.ch10.nl] acme: Preparing to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] acme: Trying to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [traefik.t.ch10.nl] Checking DNS record propagation using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] Wait [timeout: 10m0s, interval: 10s]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [portainer.t.ch10.nl] acme: Trying to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [portainer.t.ch10.nl] Checking DNS record propagation using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] Wait [timeout: 10m0s, interval: 10s]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] acme: Trying to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [pihole.t.ch10.nl] Checking DNS record propagation using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] Wait [timeout: 10m0s, interval: 10s]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [ip.t.ch10.nl] acme: Trying to solve DNS-01"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] [ip.t.ch10.nl] Checking DNS record propagation using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:15Z" level=info msg="legolog: [INFO] Wait [timeout: 10m0s, interval: 10s]"
traefik      | time="2018-12-26T14:38:16Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] acme: Trying to solve DNS-01"
traefik      | time="2018-12-26T14:38:16Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] Checking DNS record propagation using [1.1.1.1:53]"
traefik      | time="2018-12-26T14:38:16Z" level=info msg="legolog: [INFO] Wait [timeout: 10m0s, interval: 10s]"
traefik      | time="2018-12-26T14:42:02Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] The server validated our request"
traefik      | time="2018-12-26T14:42:04Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] acme: Validations succeeded; requesting certificates"
traefik      | time="2018-12-26T14:42:09Z" level=info msg="legolog: [INFO] [docs.t.ch10.nl] Server responded with a certificate."
traefik      | time="2018-12-26T14:42:09Z" level=info msg="Server configuration reloaded on :8080"
traefik      | time="2018-12-26T14:42:09Z" level=info msg="Server configuration reloaded on :80"
traefik      | time="2018-12-26T14:42:09Z" level=info msg="Server configuration reloaded on :443"```
```

The other propagation checks end up timing out (I think their TXT records got overwritten by the winner that did succeed).

I've solved my direct issue by using a wildcard certificate but I wanted to report this anyway since I do think there actually is a bug.

See also:

Possible first race condition window in `func (d *DNSProvider) Present`:

- https://github.com/xenolf/lego/blob/master/providers/dns/transip/transip.go#L94
- https://github.com/xenolf/lego/blob/master/providers/dns/transip/transip.go#L108

And possible second race condition window in `func (d *DNSProvider) CleanUp`:

- https://github.com/xenolf/lego/blob/master/providers/dns/transip/transip.go#L130
- https://github.com/xenolf/lego/blob/master/providers/dns/transip/transip.go#L144

**Comment by ldez** [comment]
Note that the issue appears only if several Host rules use the same "root domain".
A wildcard certificate is the best way to handle this case.

The [TransIP API](https://www.transip.nl/transip/api/) only provide `SetDNSEntries` method to add/update TXT records.
Due to this limitation, it will be hard to find a better solution than use a wildcard certificates.

**Comment by ronaldevers** [comment]
Yeah I agree. Only solution I can come up with is to bluntly add a `sync.Mutex` to the `transip.DNSProvider` struct and use that around both `GetInfo/SetDNSEntries` windows. It's maybe not very pretty but it is a local solution to a provider-related issue so I think that should be okay... If you're requesting only one certificate, then you won't notice it and if you are requesting multiple concurrently with the same transip.DNSProvider, then you may notice that it will be slower, but at least it would work?

I'm not sure if the `DNSProvider.Cleanup` calls would wipe out other challenges though. So there may still be a problem even with a mutex.

**Comment by traefiker** [comment]
Closed by #4376.

**Comment by mark-veenstra** [comment]
@ronaldevers Since you are located in NL and have used TransIP before, any ideas on this one: #5141
